File: richos/mobile/dev/mac-server.rs
//! Real Rust phone HTTP/auth/SSE, durable intake, Spine and gated timeline.
//! Only cognition and the Tauri AppHandle adapter are replaced. No model credentials,
//! production pairing, login keychain, microphone or desktop window are accessed.
//! Started by `mobile.mjs lab mac`; this module exists only in the Rust test binary.
use crate::phone::{
    self,
    api_base::ApiBaseDesk,
    assets::PhoneApp,
    ca::PhoneCa,
    device::DeviceDesk,
    listen::{tls_config, Listener},
    names::LocalNames,
    routes::{Accepted, Bridge, Channel, StopSwitch},
    stream::{PhoneHub, PhoneLiveEmitter},
};
use richos_core::{
    cognition::MockLeaseFactory,
    entity::{Entity, EntityId, EntityRegistry},
    ledger::Ledger,
    spine::Spine,
    steering::TurnControl,
};
use serde_json::{json, Value};
use std::{
    net::{IpAddr, Ipv4Addr},
    path::PathBuf,
    sync::{Arc, Mutex},
    time::{Duration, Instant},
};

// Test-only storage survives an isolated process restart. Production uses Keychain.
struct LabSecrets { directory: PathBuf }
impl phone::secrets::SecretStore for LabSecrets {
    fn get(&self, account: &str) -> Result<Option<Vec<u8>>, phone::PhoneError> {
        let path = self.directory.join(format!("secret-{}",phone::hex(&phone::sha256(account.as_bytes()))));
        match std::fs::read(path) { Ok(bytes)=>Ok(Some(bytes)), Err(e) if e.kind()==std::io::ErrorKind::NotFound=>Ok(None), Err(e)=>Err(e.into()) }
    }
    fn put(&self, account: &str, bytes: &[u8]) -> Result<(), phone::PhoneError> {
        use std::os::unix::fs::OpenOptionsExt;
        use std::io::Write;
        let path = self.directory.join(format!("secret-{}",phone::hex(&phone::sha256(account.as_bytes()))));
        let mut file = std::fs::OpenOptions::new().create(true).truncate(true).write(true).mode(0o600).open(path)?;
        file.write_all(bytes)?; file.sync_all()?; Ok(())
    }
    fn delete(&self, account: &str) -> Result<(), phone::PhoneError> {
        let path = self.directory.join(format!("secret-{}",phone::hex(&phone::sha256(account.as_bytes()))));
        match std::fs::remove_file(path) { Ok(())=>Ok(()), Err(e) if e.kind()==std::io::ErrorKind::NotFound=>Ok(()), Err(e)=>Err(e.into()) }
    }
}

struct IsolatedBridge {
    spine: Arc<Mutex<Spine>>,
    control: TurnControl,
    thread: String,
    entity: EntityId,
    workers: Mutex<Vec<std::thread::JoinHandle<()>>>,
}
impl Bridge for IsolatedBridge {
    fn submit_text(&self, thread_id: Option<&str>, text: &str) -> Result<Accepted, String> {
        let thread = thread_id.unwrap_or(&self.thread);
        if thread != self.thread {
            return Err("Unknown isolated conversation".into());
        }
        let record = self
            .control
            .submit_from_channel(thread, Some(self.entity.clone()), text, "phone")
            .map_err(|e| e.to_string())?;
        let spine = Arc::clone(&self.spine);
        self.workers
            .lock()
            .unwrap()
            .push(std::thread::spawn(move || {
                spine
                    .lock()
                    .unwrap()
                    .poll_intake()
                    .expect("isolated phone intake drain");
            }));
        Ok(Accepted {
            message_id: format!("intake_{}", record.id()),
            thread_id: thread.into(),
            at: phone::now_millis(),
        })
    }
    fn snapshot(&self, thread_id: Option<&str>) -> Result<Value, String> {
        crate::timeline_view::timeline_payload(
            &*self.spine.lock().unwrap(),
            thread_id.unwrap_or(&self.thread),
        )
    }
    fn current_thread(&self) -> Option<(String, String)> {
        Some((self.thread.clone(), "Isolated mobile check".into()))
    }
    fn threads(&self) -> Vec<(String, String)> {
        vec![self.current_thread().unwrap()]
    }
}

#[test]
#[ignore = "CLI-owned isolated server; requires an explicit scratch directory and stop signal"]
fn serve() {
    let dir = PathBuf::from(
        std::env::var_os("RICHOS_MOBILE_MAC_DIR").expect("CLI must supply isolated scratch"),
    );
    let dir = dir.canonicalize().unwrap();
    assert!(dir.starts_with("/Volumes/E1TB/"));
    let restarting = std::env::var("RICHOS_MOBILE_MAC_RESTART").as_deref() == Ok("1");
    if restarting {
        assert_eq!(std::fs::read_to_string(dir.join("lab-owner")).unwrap(), "richos-mobile-isolated-v1");
    } else {
        assert_eq!(std::fs::read_dir(&dir).unwrap().count(),0,"never reuse an existing data directory");
        std::fs::write(dir.join("lab-owner"),"richos-mobile-isolated-v1").unwrap();
    }
    let connect_token_file = std::env::var_os("RICHOS_MOBILE_CONNECT_TOKEN_FILE");
    let managed = connect_token_file.is_some();
    let origin = std::env::var("RICHOS_MOBILE_MAC_ORIGIN").expect("explicit public test origin");
    let names = LocalNames {
        host: "Mobile integration".into(),
        bonjour: "localhost".into(),
        addresses: vec![IpAddr::V4(Ipv4Addr::LOCALHOST)],
    };
    let ca = PhoneCa::open(&dir, &LabSecrets {directory:dir.clone()}, names).unwrap();
    let mut spine = Spine::new(Ledger::open(dir.join("ledger.jsonl")).unwrap());
    let root = dir.to_str().unwrap();
    spine.set_entity_registry(
        EntityRegistry::new(vec![
            Entity::new("mobile-check", "Mobile check", &[root]).unwrap()
        ])
        .unwrap(),
    );
    let entity = EntityId::parse("mobile-check").unwrap();
    let thread = if restarting { std::fs::read_to_string(dir.join("lab-thread")).unwrap() }
        else { let thread=spine.create_thread("Isolated mobile check",&entity).unwrap();
            std::fs::write(dir.join("lab-thread"),&thread).unwrap(); thread };
    spine.switch_thread(&thread).unwrap();
    // Empty scripted replies make the existing test provider emit `ack: <fresh message>`.
    spine.set_lease_factory(Box::new(MockLeaseFactory::new(vec![])));
    let control = TurnControl::open(dir.join("intake.jsonl")).unwrap();
    spine.set_turn_control(control.clone());
    let hub = PhoneHub::new();
    spine.set_live_observer(Box::new(PhoneLiveEmitter::new(Arc::clone(&hub))));
    let bridge = Arc::new(IsolatedBridge {
        spine: Arc::new(Mutex::new(spine)),
        control,
        thread,
        entity,
        workers: Mutex::new(vec![]),
    });
    let devices = Arc::new(DeviceDesk::open(&dir).unwrap());
    let channel = Arc::new(Channel {
        devices: Arc::clone(&devices),
        rejected: StopSwitch::unwired(),
        api_base: Arc::new(ApiBaseDesk::only(&origin)),
        hub,
        bridge: bridge.clone(),
        assets: PhoneApp::embedded(),
        vapid_public: phone::push::VapidKey::generate()
            .unwrap()
            .application_server_key(),
        fingerprint_hex: ca.fingerprint_hex(),
        pairing_path: Mutex::new(if managed { phone::device::PairedVia::CONNECT } else { phone::device::PairedVia::TAILNET }),
    });
    let mut listener = if managed {
        Listener::start_connect(channel, phone::connect::PORT).unwrap()
    } else {
        Listener::start(channel, tls_config(&ca.leaf_der,&ca.leaf_key_pkcs8).unwrap(), &[IpAddr::V4(Ipv4Addr::LOCALHOST)],0).unwrap()
    };
    let _connector = connect_token_file.map(|file| {
        let path = PathBuf::from(file).canonicalize().unwrap();
        assert!(path.starts_with("/Volumes/E1TB/"));
        let token = std::fs::read_to_string(path).unwrap();
        let helper = std::env::var_os("RICHOS_CONNECT_HELPER").expect("explicit test helper");
        phone::connect::supervisor::Supervisor::start(&PathBuf::from(helper),token).unwrap()
    });
    if !devices.is_paired() { devices.open_pairing().unwrap(); }
    let ready = json!({"protocol":if managed {"http"} else {"https"},"port":listener.bound[0].port(), "ca":dir.join("phone/ca.crt"),
        "pairLink":devices.pairing_window().map(|window|format!("{origin}/#pair={}",window.code)),
        "words":ca.fingerprint_words().join(" "), "replyMarker":"ack: ", "pid":std::process::id()});
    std::fs::write(dir.join("ready.new"), ready.to_string()).unwrap();
    std::fs::rename(dir.join("ready.new"), dir.join("ready.json")).unwrap();
    let deadline = Instant::now() + Duration::from_secs(3600);
    while !dir.join("stop").exists() && Instant::now() < deadline {
        if let Ok(payload) = bridge.snapshot(None) {
            std::fs::write(dir.join("timeline.new"), payload.to_string()).unwrap();
            std::fs::rename(dir.join("timeline.new"), dir.join("timeline.json")).unwrap();
        }
        std::thread::sleep(Duration::from_millis(200));
    }
    listener.stop();
    for worker in bridge.workers.lock().unwrap().drain(..) {
        worker.join().unwrap();
    }
    std::fs::write(
        dir.join("timeline.json"),
        bridge.snapshot(None).unwrap().to_string(),
    )
    .unwrap();
}

File: richos/mobile/dev/mac-server.mjs
// Launch the production Rust phone stack inside its test binary, with isolated state.
import { spawn } from 'node:child_process';
import { randomUUID } from 'node:crypto';
import { once } from 'node:events';
import { createRequire } from 'node:module';
import { dirname, join, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';
import { existsSync, readFileSync, writeFileSync, rmSync, mkdirSync, openSync, closeSync, copyFileSync } from 'node:fs';
import http from 'node:http';
import https from 'node:https';
const require = createRequire(import.meta.url);
const { createScratch } = require('../test/storage.cjs');
const root = resolve(dirname(fileURLToPath(import.meta.url)), '../..');
const pause = ms => new Promise(resolve => setTimeout(resolve, ms));

export async function startMac(cache) {
  const origin = process.env.RICHOS_MOBILE_MAC_ORIGIN || 'https://localhost';
  if (new URL(origin).origin !== origin || !origin.startsWith('https://')) throw Error('Mac test origin must be an HTTPS origin');
  rmSync(join(cache, 'mac.json'), { force: true });
  const scratch = createScratch('rust-server');
  const data = join(scratch, 'data'); mkdirSync(data, { mode: 0o700 });
  const env = { ...process.env, TMPDIR: scratch + '/', RICHOS_MOBILE_MAC_DIR: data, RICHOS_MOBILE_MAC_ORIGIN: origin };
  const log = join(cache, 'mac-server.log');
  const fd = openSync(log, 'w', 0o600);
  let child, stopped, proxy, serving = false, binary;
  function killOwned(signal) {
    if (!child?.pid) return;
    try { process.kill(-child.pid, signal); } catch (error) { if (error.code !== 'ESRCH') throw error; }
  }
  let closed = false;
  async function close() {
    if (closed) return; closed = true;
    proxy?.closeAllConnections(); proxy?.close();
    writeFileSync(join(data, 'stop'), 'stop');
    if (child && child.exitCode === null) {
      let timeout;
      const finished = await Promise.race([stopped.then(() => true), new Promise(resolve => { timeout = setTimeout(() => resolve(false), 5000); })]);
      clearTimeout(timeout);
      if (!finished) {
        killOwned('SIGTERM');
        const force = setTimeout(() => killOwned('SIGKILL'), 2000);
        try { await stopped; } finally { clearTimeout(force); }
      }
    }
    for (const name of ['timeline.json', 'ledger.jsonl', 'intake.jsonl']) {
      if (existsSync(join(data, name))) copyFileSync(join(data, name), join(cache, 'mac-' + name));
    }
    closeSync(fd); rmSync(scratch, { recursive: true, force: true });
    if (serving && child.exitCode !== 0) throw Error(`Rust server did not shut down cleanly. Read ${log}`);
  }
  try {
    const cargo = existsSync(join(process.env.HOME, '.cargo/bin/cargo')) ? join(process.env.HOME, '.cargo/bin/cargo') : 'cargo';
    child = spawn(cargo, ['test', '--manifest-path', join(root, 'app/src-tauri/Cargo.toml'), '--no-run', '--message-format=json'], { env, detached: true, stdio: ['ignore', 'pipe', fd] });
    stopped = once(child, 'close');
    let compiled = ''; child.stdout.on('data', chunk => { compiled += chunk; });
    let force;
    const timeout = setTimeout(() => { killOwned('SIGTERM'); force = setTimeout(() => killOwned('SIGKILL'), 2000); }, 300000);
    let code;
    try { [code] = await stopped; } finally { clearTimeout(timeout); clearTimeout(force); }
    writeFileSync(join(cache, 'mac-build.jsonl'), compiled);
    if (code !== 0) throw Error(`Rust server build failed. Read ${log} and mac-build.jsonl`);
    binary = compiled.split('\n').filter(Boolean).map(line => JSON.parse(line))
      .find(row => row.reason === 'compiler-artifact' && row.profile.test && row.target.name === 'richos-tauri' && row.executable)?.executable;
    if (!binary) throw Error('Cargo did not produce the Rust test executable');
    child = spawn(binary, ['mobile_mac_server::serve', '--exact', '--ignored', '--nocapture'], { env, detached: true, stdio: ['ignore', fd, fd] });
    stopped = once(child, 'close');
    const deadline = Date.now() + (process.env.RICHOS_MOBILE_VOICE_TEST === '1' ? 120000 : 20000);
    while (!existsSync(join(data, 'ready.json'))) {
      if (child.exitCode !== null || Date.now() > deadline) throw Error(`Rust server did not start. Read ${log}`);
      await pause(50);
    }
    serving = true;
    const state = JSON.parse(readFileSync(join(data, 'ready.json'), 'utf8'));
    const ca = readFileSync(state.ca);
    if (state.protocol !== "http") {
    proxy = http.createServer((request, response) => {
      const upstream = https.request({ hostname: '127.0.0.1', servername: 'localhost', port: state.port,
        path: request.url, method: request.method, headers: request.headers, ca }, result => {
        response.writeHead(result.statusCode, result.headers); response.flushHeaders(); result.pipe(response);
      });
      upstream.on('error', () => { if (!response.headersSent) response.writeHead(502); response.end(); });
      request.pipe(upstream); response.on('close', () => upstream.destroy());
    });
    await new Promise((resolve, reject) => { proxy.once('error', reject); proxy.listen(0, '127.0.0.1', resolve); });
    }
    const config = { pairLink: state.pairLink, words: state.words, replyMarker: state.replyMarker, serverRun: randomUUID() };
    writeFileSync(join(cache, 'mac-test-config.json'), JSON.stringify(config), { mode: 0o600 });
    const publicState = { ...state, port: proxy ? proxy.address().port : state.port, rustPort: state.port, origin, log, data };
    writeFileSync(join(cache, 'mac.json'), JSON.stringify(publicState, null, 2), { mode: 0o600 });
    async function restart() {
      if (closed || !serving) throw Error('Only the owned running Mac fixture can restart');
      child.kill('SIGKILL'); // One exact owned process. Its guard must notice the parent pipe closing.
      await stopped;
      rmSync(join(data,'ready.json'),{force:true});
      rmSync(join(data,'stop'),{force:true});
      child = spawn(binary, ['mobile_mac_server::serve','--exact','--ignored','--nocapture'], {
        env: {...env,RICHOS_MOBILE_MAC_RESTART:'1'}, detached:true, stdio:['ignore',fd,fd] });
      stopped = once(child,'close');
      const deadline = Date.now() + (process.env.RICHOS_MOBILE_VOICE_TEST === '1' ? 120000 : 20000);
      while (!existsSync(join(data,'ready.json'))) {
        if (child.exitCode !== null || Date.now() > deadline) throw Error(`Restart failed. Read ${log}`);
        await pause(50);
      }
      Object.assign(state,JSON.parse(readFileSync(join(data,'ready.json'),'utf8')));
      publicState.pid = state.pid; publicState.rustPort = state.port;
      writeFileSync(join(cache,'mac.json'),JSON.stringify(publicState),{mode:0o600});
      return publicState;
    }
    return { state: publicState, close, restart, get exited() { return stopped; } };
  } catch (error) { await close(); throw error; }
}
export async function serveMac(cache) {
  const server = await startMac(cache);
  console.log(JSON.stringify({ ok: true, mode: 'mac', state: server.state }));
  let timer, done;
  const requested = new Promise(resolve => {
    done = () => resolve('requested');
    timer = setTimeout(done, 55 * 60 * 1000);
    process.once('SIGINT', done); process.once('SIGTERM', done);
  });
  try {
    const reason = await Promise.race([requested, server.exited.then(() => 'exited')]);
    if (reason === 'exited') throw Error(`Rust server exited unexpectedly. Read ${server.state.log}`);
  } finally {
    clearTimeout(timer); process.off('SIGINT', done); process.off('SIGTERM', done);
    await server.close();
  }
  return { stopped: true };
}
